File: android/src/main/kotlin/id/my/burganbank/flutter_shield/EncryptionStrategy.kt
package id.my.burganbank.flutter_shield

import java.security.KeyPair
import android.content.Context
import id.my.burganbank.flutter_shield.model.*

interface EncryptionStrategy {

    @Throws(Exception::class)
    fun storeCertificate(tag: String, certificateData: ByteArray): Boolean

    @Throws(Exception::class)
    fun getCertificate(tag: String): String?

    fun removeCertificate(tag: String): Boolean
    
    @Throws(Exception::class)
    fun storeServerPrivateKey(tag: String, privateKeyData: ByteArray): Boolean

    @Throws(Exception::class)
    fun getServerKey(tag: String): String?

    @Throws(Exception::class)
    fun generateKeyPair(accessControlParam: AccessControlParam): KeyPair

    fun removeKey(tag: String, flag: String): Boolean

    @Throws(Exception::class)
    fun isKeyCreated(tag: String, flag: String): Boolean?

    @Throws(Exception::class)
    fun getPublicKey(tag: String): String?

    @Throws(Exception::class)
    fun encrypt(message: String, tag: String): ByteArray?

    @Throws(Exception::class)
    fun decrypt(message: ByteArray, tag: String): String?

    @Throws(Exception::class)
    fun sign(tag: String, message: ByteArray): String?

    @Throws(Exception::class)
    fun verify(tag: String, plainText: String, signature: String): Boolean

    @Throws(Exception::class)
    fun decryptWithAES(encryptedData: ByteArray, aesKey: ByteArray): String?
}

File: ios/Classes/EncryptionStrategy.swift
import Foundation
import LocalAuthentication

// Abstraction/Protocol class of EncryptionStrategy
protocol EncryptionStrategy {
    
    func storeCertificate(certificateData: Data, tag: String) throws -> Bool

    func getCertificate(tag: String) throws -> String?

    func removeCertificate(tag: String) throws -> Bool
    
    func storeServerPrivateKey(privateKeyData: Data, tag: String) throws -> Bool

    func getServerKey(tag: String) throws -> String?
    
    func generateKeyPair(accessControlParam: AccessControlParam) throws -> SecKey
    
    func removeKey(tag: String, flag: String) throws -> Bool
    
    func isKeyCreated(tag: String, flag: String) throws -> Bool?
    
    func getPublicKey(tag: String) throws -> String?
    
    func encrypt(message: String, tag: String) throws -> FlutterStandardTypedData?
    
    func decrypt(message: Data, tag: String) throws -> String?
    
    func sign(tag: String, message: Data) throws -> String?
    
    func verify(tag: String, plainText: String, signature: String) throws -> Bool
}
